src/magic_folder/magic_folder.py

```diff
@@ -838,3 +852,141 @@ def add_file(self, path):
+    information as some local snapshots that already exist.
+    """
+    def upload_local_snapshots():
+        """
+        Find local uncommitted snapshots and commit them to the grid.
+
+        :return Deferred[None]: A Deferred that fires when an effort has been
+            made to commit at least some local uncommitted snapshots.
+        """
+
+
+@implementer(IRemoteSnapshotCreator)
+@attr.s
+class RemoteSnapshotCreator(object):
+    _state_db = attr.ib()
+    _local_author = attr.ib()
+    _tahoe_client = attr.ib()
+    _upload_dircap = attr.ib()
+
+    @eliotutil.inline_callbacks
+    def upload_local_snapshots(self):
+        """
+        Check the db for uncommitted LocalSnapshots, deserialize them from the on-disk
+        format to LocalSnapshot objects and commit them into the grid.
+        """
+
+        # get the mangled paths for the LocalSnapshot objects in the db
+        localsnapshot_names = self._state_db.get_all_localsnapshot_paths()
+
+        # XXX: processing this table should be atomic. i.e. While the upload is
+        # in progress, a new snapshot can be created on a file we already uploaded
+        # but not removed from the db and if it gets removed from the table later,
+        # the new snapshot gets lost. Perhaps this can be solved by storing each
+        # LocalSnapshot in its own row than storing everything in a blob?
+        # https://github.com/LeastAuthority/magic-folder/issues/197
+        for name in localsnapshot_names:
+            action = UPLOADER_SERVICE_UPLOAD_LOCAL_SNAPSHOTS(relpath=name)
+            try:
+                with action:
+                    yield self._upload_one_snapshot(name)
+            except:
+                # Unable to reach Tahoe storage nodes because of network
+                # errors or because the tahoe storage nodes are
+                # offline. Retry?
+                pass
+
+    @eliotutil.inline_callbacks
+    def _upload_one_snapshot(self, name):
+        """
+        Upload all of the snapshots for a particular path.
+        """
+        # deserialize into LocalSnapshot
+        snapshot = self._state_db.get_local_snapshot(name)
+
+        remote_snapshot = yield write_snapshot_to_tahoe(
+            snapshot,
+            self._local_author,
+            self._tahoe_client,
+        )
+
+        # At this point, remote snapshot creation successful for
+        # the given relpath.
+        # store the remote snapshot capability in the db.
+        yield self._state_db.store_remotesnapshot(name, remote_snapshot)
+
+        # update the entry in the DMD
+        yield self._tahoe_client.add_entry_to_mutable_directory(
+            self._upload_dircap,
+            magicpath.path2magic(name),
+            remote_snapshot.capability.encode('utf-8'),
+            replace=True,
+        )
+
+        # Remove the local snapshot content from the stash area.
+        snapshot.content_path.remove()
```

Comment:
I'm trying to think of failure cases, and I think we might want to mostly-ignore errors here yet still remove the database entry? (Thinking: if there's some error removing the content, could it be partially-removed at that point? If so, then the database is kind of inconsistent: there's a database entry, but corrupted or missing "content"). Caveat: **is** it possible that the `remove` fails such that the content is gone/missing but still produces an error?

Reply:
> I'm trying to think of failure cases, and I think we might want to mostly-ignore errors here yet still remove the database entry?

Something like this, probably.  There are a bunch of atomicity issues with this function as written now but they're not easily fixable with the current database schema.  I'm tempted to leave this for now and make another pass later on.  It will work in the happy-case, at least ...

> Caveat: is it possible that the remove fails such that the content is gone/missing but still produces an error?

I think the rule of thumb for answering "Is it possible for a filesystem to have confounding behavior X?" is "yes".